// office-ui-fabric-react/lib/demo/pages/ChoiceGroupPage/examples/ChoiceGroup.Basic.Example.js
"use strict";
var __extends = (this && this.__extends) || function (d, b) {
    for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p];
    function __() { this.constructor = d; }
    d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
};
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
var React = require("react");
var index_1 = require("../../../../index");
var ChoiceGroupBasicExample = (function (_super) {
    __extends(ChoiceGroupBasicExample, _super);
    function ChoiceGroupBasicExample() {
        var _this = _super.call(this) || this;
        _this.state = {
            imageKey: ''
        };
        _this._onImageChoiceGroupChange = _this._onImageChoiceGroupChange.bind(_this);
        return _this;
    }
    ChoiceGroupBasicExample.prototype.render = function () {
        return (React.createElement("div", null,
            React.createElement(index_1.ChoiceGroup, { options: [
                    {
                        key: 'A',
                        text: 'Option A'
                    },
                    {
                        key: 'B',
                        text: 'Option B',
                        checked: true
                    },
                    {
                        key: 'C',
                        text: 'Option C',
                        disabled: true
                    },
                    {
                        key: 'D',
                        text: 'Option D',
                        checked: true,
                        disabled: true
                    }
                ], onChange: this._onChange, label: 'Pick one', required: true }),
            React.createElement(index_1.ChoiceGroup, { label: 'Pick one image', options: [
                    {
                        key: 'bar',
                        checked: this.state.imageKey === 'bar',
                        imageSrc: 'dist/choicegroup-bar-unselected.png',
                        selectedImageSrc: 'dist/choicegroup-bar-selected.png',
                        imageSize: { width: 50, height: 50 },
                        text: 'Bar chart'
                    },
                    {
                        key: 'pie',
                        checked: this.state.imageKey === 'pie',
                        imageSrc: 'dist/choicegroup-pie-unselected.png',
                        selectedImageSrc: 'dist/choicegroup-pie-selected.png',
                        imageSize: { width: 50, height: 50 },
                        text: 'Pie chart'
                    }
                ], onChange: this._onImageChoiceGroupChange })));
    };
    ChoiceGroupBasicExample.prototype._onChange = function (ev, option) {
        console.dir(option);
    };
    ChoiceGroupBasicExample.prototype._onImageChoiceGroupChange = function (ev, option) {
        this.setState({
            imageKey: option.key
        });
    };
    return ChoiceGroupBasicExample;
}(React.Component));
__decorate([
    index_1.autobind
], ChoiceGroupBasicExample.prototype, "_onChange", null);
exports.ChoiceGroupBasicExample = ChoiceGroupBasicExample;
